### PR TITLE
[SPARK-19089][SQL] Add support for nested sequences

### PR DESCRIPTION
**Pull request by michalsenkyr**
## What changes were proposed in this pull request?

Replaced specific sequence encoders with generic sequence encoder to enable nesting of sequences.

Does not add support for nested arrays as that cannot be solved in this way.

## How was this patch tested?

```bash
build/mvn -DskipTests clean package && dev/run-tests
```

Additionally in Spark shell:

```
scala> Seq(Seq(Seq(1))).toDS.collect()
res0: Array[Seq[Seq[Int]]] = Array(List(List(1)))
```

### CODE REVIEW

**Comment by cloud-fan** [comment]
ok to test

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetPrimitiveSuite.scala [hunk 1]
```diff
@@ -258,6 +258,10 @@ class DatasetPrimitiveSuite extends QueryTest with SharedSQLContext {
       ListClass(List(1)) -> Queue("test" -> SeqClass(Seq(2))))
   }
 
+  test("nested sequences") {
+    checkDataset(Seq(Seq(Seq(1))).toDS(), Seq(Seq(1)))
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetPrimitiveSuite.scala]: let's also add test for specific collection type, e.g. `List(Queue(1))`

**Comment by cloud-fan** [comment]
LGTM

**Comment by gatorsmile** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master/2.2!

**Pull request closed without merging**